packages/core/src/ui/Base/TwitterEmbed.tsx

```diff
@@ -0,0 +1,124 @@
+import { useEffect, useCallback, useState, useRef, RefObject } from "react";
+
+const twitterWidgetJs = "https://platform.twitter.com/widgets.js";
+const message = "Loading tweet...";
+
+interface TweetConfig {
+  theme: string;
+}
+
+type TweetState = {
+  isLoading: boolean;
+  message: string | null;
+};
+
+declare global {
+  interface Window {
+    twttr: {
+      widgets: {
+        createTweet: (
+          id: string,
+          ref: RefObject<HTMLDivElement>,
+          options: TweetConfig
+        ) => Promise<any>; // TODO type
+        load: (ref: RefObject<HTMLDivElement>) => void;
+      };
+    };
+  }
+}
+
+export default function TwitterEmbed({ url, ...props }) {
+  const ref = useRef<HTMLDivElement | null>(null);
+  const [state, setState] = useState<TweetState>({
+    isLoading: true,
+    message: message,
+  });
+
+  const tweetId = url.split("status/").pop();
+
+  const renderTweet = useCallback(() => {
+    if (typeof window !== undefined) {
+      window.twttr.widgets
+        .createTweet(tweetId, ref.current as any, {
+          theme: "dark",
+        })
+        .then((el) => {
+          if (el) return setState((prev) => ({ ...prev, isLoading: false }));
+          return setState((prev) => ({
+            ...prev,
+            message: null,
+          }));
+        });
+      return window.twttr.widgets.load(ref.current as any);
+    }
+  }, [tweetId]);
+
+  useEffect(() => {
+    let componentMounted = true;
+
+    const script = document.createElement("script");
+    script.src = twitterWidgetJs;
+    script.async = true;
+    script.onload = () => renderTweet();
+
+    if (componentMounted) {
+      if (!window.twttr) {
+        document.head.appendChild(script);
+      } else {
+        renderTweet();
+      }
+    }
+
+    return () => {
+      setState({ isLoading: true, message: message });
+      componentMounted = false;
+    };
+  }, [renderTweet]);
+
+  return (
+    <>
+      {state.isLoading && state.message && (
```

Comment:
This doesn't really make sense when I'm reading this. It's not self-explanatory. Why would a message affect displaying a placeholder? Or why the placeholder doesn't need to be displayed if there is no message, even if the state is still loading?